lib/clib/Corres_UL_C.thy

```diff
@@ -137,6 +141,15 @@ lemma ccorresI':
   apply simp
   done
 
+lemma ccorresI_empty_handler_stack:
+  assumes rl:
+    "\<And>s s' xstate.
+      \<lbrakk>(s, s') \<in> srel; G s; s' \<in> G'; \<Gamma> \<turnstile> \<langle>c, Normal s'\<rangle> \<Rightarrow> xstate; \<not> snd (a s)\<rbrakk>
+      \<Longrightarrow> (\<exists>t'. xstate = Normal t' \<and> (\<exists>(r, t) \<in> fst (a s). (t, t') \<in> srel \<and> rrel r (xf t')))"
+  shows "ccorresG srel \<Gamma> rrel xf G G' [] a c"
```

Comment:
This is strange. I think maybe the name should be tweaked to `ccorres_empty_handler_stackI` so people don't look for a `ccorresI` like the `ccorresG`.
The other thing is, do you have a small comment on when one would use this? I'm staring at it and can't figure it out.

Reply:
It is a bit confusing having `ccorresG` as an abbreviation and then having `ccorresI` and `ccorresE` (probably as well as others) being intro/elim style rules. I'm happy to change the name, but to me this name makes more sense because it is an intro rule for (ultimately a) `ccorres_underlying` statement, but with some extra restrictions, namely there being an empty handler stack.

As for when to use it, I'm no expert in C semantics or CRefine, so the handler stack is still a bit mysterious to me. Gerwin mentioned on Mattermost that if we have a schematic handler stack, then to prove `ccorres`, we would need to make assumptions about what the handler stack does, probably something like it taking any `Abrupt` state to a `Normal` state. So if we didn't want to do something like that and we still wanted to prove a `ccorres` rule, we would need to show that the `c` function in this rule always returns a `Normal` state, in which case, the handler stack may as well be empty, since it won't be used. Unfolding `ccorres_underlying_def` will leave us with terms involving `exec_handlers` which aren't as nice to deal with as regular `hoarep` statements, and the `exec_handler` statement doesn't mean much with an empty handler stack anyway.

So maybe a comment to help explain how it could be used could say "for proving low-level `ccorresG` statements where the concrete function always returns a `Normal` state, and therefore, the handler stack will not be used, and can be considered empty". But I'm not sure that's more helpful than what someone might glean from the name of the lemma

Reply:
Yes please to this comment. Even if it's a skill issue, I find your explanation and comment helpful.